theano/tensor/basic.py

```diff
@@ -5902,24 +5902,74 @@ class Diagonal(Op):
 
     """
     __props__ = ("offset", "axis1", "axis2")
+    default_offset = 0
+    default_axis1 = 0
+    default_axis2 = 1
 
     def __init__(self, offset=0, axis1=0, axis2=1):
-        if numpy_diagonal_return_view:
+        self.view = view
+        if self.view and not numpy_diagonal_return_view:
+            warnings.warn("View will forced to False. Diagonal property view is "
+                          "set to True but numpy version %s and prior versions of "
+                          "numpy.diagonal() do not return a view. Update "
+                          "numpy to use Diagonal(view=True)" % 
+                          numpy.version.version)
+            self.view = False
+        if self.view:
             self.view_map = {0: [0]}
         self.offset = offset
         self.axis1 = axis1
         self.axis2 = axis2
 
-    def make_node(self, x):
-        x = as_tensor_variable(x)
-        assert x.ndim >= 2
-        return Apply(self, [x], [tensor(dtype=x.dtype,
-                                        broadcastable=[False] * (x.ndim - 1))])
+    def make_node(self, _x):
+        if not isinstance(_x, theano.Variable):
+            x = as_tensor_variable(_x)
+        else:
+            x = _x
+
+        if x.ndim < 2:
+            raise ValueError('Diagonal needs an input with 2 or more '
+                             'dimensions', x)
+        return Apply(self, [x], [x.type.__class__(
+            dtype=x.dtype,
+            broadcastable=[False] * (x.ndim - 1))()])
 
     def perform(self, node, inputs, outputs):
         (x,) = inputs
         (z,) = outputs
-        z[0] = x.diagonal(self.offset, self.axis1, self.axis2)
+        # zero-dimensional matrices ...
```

Comment:
Historically NumPy diagonal was always returning a copy.
Le 23 avr. 2016 11:51, "Xavier Bouthillier" <notifications@github.com> a
écrit :

> In theano/tensor/basic.py
> <https://github.com/Theano/Theano/pull/4424#discussion_r60830445>:
>
> >
> >      def perform(self, node, inputs, outputs):
> >          (x,) = inputs
> >          (z,) = outputs
> > -        z[0] = x.diagonal(self.offset, self.axis1, self.axis2)
> > +        # zero-dimensional matrices ...
>
> Just don't forget to do rval.copy() if we use numpy.diagonal and self.view
> is False
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/4424/files/c1a864ccf486d74df2625c15a67ded78c5b85cdf#r60830445>
>